pyccl/pk2d.py

```diff
@@ -527,3 +533,45 @@ def parse_pk2d(cosmo, p_of_k_a, is_linear=False):
             pk = cosmo.get_nonlin_power(name)
         psp = pk.psp
     return psp
+
+
+def sample_pk_function(pkfunc, cosmo=None):
+    """ Samples a function of wavenumber and scale factor at the default
+    internal nodes used by CCL.
+
+    Args:
+        pkfunc : :obj:`callable`
+            Function with signature ``f(k, a)`` which takes vectorized input
+            in ``k`` (wavenumber in :math:`\\mathrm{Mpc}^{-1}``) and a scale
+            factor`a``, and returns the value of the corresponding quantity.
+            ``pkfunc`` will be sampled at the values of ``k`` and ``a`` used
+            internally by CCL to store the linear and non-linear power spectra.
+        cosmo : :class:`~pyccl.core.Cosmology`, optional cosmological
+            parameters. Used to determine sampling rates in scale factor and
+            wavenumber.
+
+    Returns:
+        a : array_like
+            Array containing the nodes in scale factor at which the power
+            spectrum has been sampled.
+        logk : array_like
+            Array containing the nodes in log(k) (natural logarithm) at which
+            the power spectrum has been sampled.
+        pk : array_like
+            Two-dimensional array of shape `[len(a), len(logk)]` containing
+            the power spectrum sampled at the corresponding values of `a` and
+            `logk` (i.e. `pk[i, j] = pkfunc(exp(logk[j]), a[i])`).
+    """
+    # Check that the input function has the right signature
+    try:
+        pkfunc(k=np.array([1E-2, 2E-2]), a=0.5)
+    except Exception:
+        raise ValueError("Can't use input function")
+
+    # Set k and a sampling from CCL parameters
+    a_arr = get_pk_spline_a(cosmo=cosmo)
+    lk_arr = get_pk_spline_lk(cosmo=cosmo)
+
+    # Compute power spectrum on 2D grid
+    pk_out = np.array([pkfunc(k=np.exp(lk_arr), a=a) for a in a_arr])
+    return a_arr, lk_arr, pk_out
```

Comment:
I know we are doing thing because I proposed it, but I just realized that, contrary to what I mentioned on Slack, we don't in fact need to pass both `pkfunc` and `cosmo`. With #918 we make `cosmo` optional and sampling is taken from the current `ccl.spline_params` state. I was the one to implement it, and I didn't remember it!

So here's an idea. Why don't we allow `pkfunc` for whoever wants to use it, but deprecate `cosmo` instead. If anyone wants to do nitpicking wrt sampling, they can just change `spline_params`. I think this way is simpler and causes the least disturbance in v3.

And there's no need to keep any of the try-except checking either. The function is used directly below that line. If it's wrong, it will exit with a useful message.

Reply:
made into classmethod